# 尼日利亚观众音乐品味的聚类分析

In [2]:
# 数据分析常用库
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import sklearn as sk
# 设置中文字体
plt.rcParams['font.sans-serif'] = ['SimHei']
plt.rcParams['axes.unicode_minus'] = False
# 去除警告
import warnings
warnings.filterwarnings('ignore')

# 读取数据
data = pd.read_csv('nigerian-songs.csv')
data

name                         album  \
0                      Sparky            Mandy & The Jungle   
1                  shuga rush  EVERYTHING YOU HEARD IS TRUE   
2                       LITT!                         LITT!   
3    Confident / Feeling Cool               Enjoy Your Life   
4                  wanted you                         rare.   
..                        ...                           ...   
525                   Kwaroro                       Kwaroro   
526             Concert Party                   Afro Escape   
527                      Loko                    Applaudise   
528                      Gift                    Applaudise   
529                    Macoma                    Applaudise   

                  artist artist_top_genre  release_date  length  popularity  \
0          Cruel Santino  alternative r&b          2019  144000          48   
1    Odunsi (The Engine)          afropop          2020   89488          30   
2                   AYLØ        indie r&b          2018  207758          40   
3             Lady Donli     nigerian pop          2019  175135          14   
4    Odunsi (The Engine)          afropop          2018  152049          25   
..                   ...              ...           ...     ...         ...   
525           J. Martins   afro dancehall          2015  193960          10   
526      Various Artists   afro dancehall          2015  252708           0   
527               Iyanya   afro dancehall          2015  223159           0   
528               Iyanya   afro dancehall          2015  278280           0   
529               Iyanya   afro dancehall          2015  213426           0   

     danceability  acousticness  energy  instrumentalness  liveness  loudness  \
0           0.666        0.8510   0.420          0.534000    0.1100    -6.699   
1           0.710        0.0822   0.683          0.000169    0.1010    -5.640   
2           0.836        0.2720   0.564          0.000537    0.1100    -7.127   
3           0.894        0.7980   0.611          0.000187    0.0964    -4.961   
4           0.702        0.1160   0.833          0.910000    0.3480    -6.044   
..            ...           ...     ...               ...       ...       ...   
525         0.735        0.6320   0.918          0.193000    0.0718    -2.582   
526         0.602        0.1770   0.883          0.000004    0.1190    -6.183   
527         0.729        0.2380   0.864          0.000000    0.1950    -3.942   
528         0.809        0.0490   0.806          0.000000    0.0659    -6.547   
529         0.816        0.1140   0.961          0.008500    0.1370    -2.204   

     speechiness    tempo  time_signature  
0         0.0829  133.015               5  
1         0.3600  129.993               3  
2         0.0424  130.005               4  
3         0.1130  111.087               4  
4         0.0447  105.115               4  
..           ...      ...             ...  
525       0.0355  127.954               4  
526       0.0778  154.010               4  
527       0.0668  121.058               4  
528       0.2340  123.997               4  
529       0.1390  106.972               4  

[530 rows x 16 columns]

In [3]:
# 查看数据的基本信息
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 530 entries, 0 to 529
Data columns (total 16 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   name              530 non-null    object 
 1   album             530 non-null    object 
 2   artist            530 non-null    object 
 3   artist_top_genre  530 non-null    object 
 4   release_date      530 non-null    int64  
 5   length            530 non-null    int64  
 6   popularity        530 non-null    int64  
 7   danceability      530 non-null    float64
 8   acousticness      530 non-null    float64
 9   energy            530 non-null    float64
 10  instrumentalness  530 non-null    float64
 11  liveness          530 non-null    float64
 12  loudness          530 non-null    float64
 13  speechiness       530 non-null    float64
 14  tempo             530 non-null    float64
 15  time_signature    530 non-null    int64  
dtypes: float64(8), int64(4), object(4)
memory us

歌曲名称、专辑、艺术家、艺术家顶级流派、发行日期、长度、流行度、舞蹈性、音乐性、能量、乐器性、活跃度、响度、语速、节奏和时值  
**各列都是530，均不存在空值，故无需进行缺失值处理。数据类型见Dtype。通过观察，这是一个比较规整的数据集，无需进行缺失值等处理。**


In [4]:
# 替换列名称，便于理解
# 原始列名: ['name', 'album', 'artist', 'artist_top_genre', 'release_date', 'length', 'popularity', 'danceability', 'acousticness', 'energy', 'instrumentalness', 'liveness', 'loudness', 'speechiness', 'tempo', 'time_signature']
data.columns = ['歌曲名称', '专辑', '歌手', '歌手主要流派', '发行日期', '时长', '流行度', '舞蹈性', '声学性', '能量值', '器乐性', '现场感', '响度(dB)', '语音性/说唱成分', '节拍速度(BPM)', '拍号']
# 查看数据的前5行
data

歌曲名称                            专辑  \
0                      Sparky            Mandy & The Jungle   
1                  shuga rush  EVERYTHING YOU HEARD IS TRUE   
2                       LITT!                         LITT!   
3    Confident / Feeling Cool               Enjoy Your Life   
4                  wanted you                         rare.   
..                        ...                           ...   
525                   Kwaroro                       Kwaroro   
526             Concert Party                   Afro Escape   
527                      Loko                    Applaudise   
528                      Gift                    Applaudise   
529                    Macoma                    Applaudise   

                      歌手           歌手主要流派  发行日期      时长  流行度    舞蹈性     声学性  \
0          Cruel Santino  alternative r&b  2019  144000   48  0.666  0.8510   
1    Odunsi (The Engine)          afropop  2020   89488   30  0.710  0.0822   
2                   AYLØ        indie r&b  2018  207758   40  0.836  0.2720   
3             Lady Donli     nigerian pop  2019  175135   14  0.894  0.7980   
4    Odunsi (The Engine)          afropop  2018  152049   25  0.702  0.1160   
..                   ...              ...   ...     ...  ...    ...     ...   
525           J. Martins   afro dancehall  2015  193960   10  0.735  0.6320   
526      Various Artists   afro dancehall  2015  252708    0  0.602  0.1770   
527               Iyanya   afro dancehall  2015  223159    0  0.729  0.2380   
528               Iyanya   afro dancehall  2015  278280    0  0.809  0.0490   
529               Iyanya   afro dancehall  2015  213426    0  0.816  0.1140   

       能量值       器乐性     现场感  响度(dB)  语音性/说唱成分  节拍速度(BPM)  拍号  
0    0.420  0.534000  0.1100  -6.699    0.0829    133.015   5  
1    0.683  0.000169  0.1010  -5.640    0.3600    129.993   3  
2    0.564  0.000537  0.1100  -7.127    0.0424    130.005   4  
3    0.611  0.000187  0.0964  -4.961    0.1130    111.087   4  
4    0.833  0.910000  0.3480  -6.044    0.0447    105.115   4  
..     ...       ...     ...     ...       ...        ...  ..  
525  0.918  0.193000  0.0718  -2.582    0.0355    127.954   4  
526  0.883  0.000004  0.1190  -6.183    0.0778    154.010   4  
527  0.864  0.000000  0.1950  -3.942    0.0668    121.058   4  
528  0.806  0.000000  0.0659  -6.547    0.2340    123.997   4  
529  0.961  0.008500  0.1370  -2.204    0.1390    106.972   4  

[530 rows x 16 columns]

#  对歌手进行分类，通过如，热度情况、发歌量、首歌年份、存续年份、等对歌手划分不同类别，帮助筛选当下有发展前景的歌手群体

In [5]:
# 根据歌手和发行日期对数据进行汇总
data_summary = data.groupby(['歌手', '发行日期']).agg({'歌曲名称': 'count', '流行度': 'sum'}).reset_index()
data_summary

歌手  发行日期  歌曲名称  流行度
0          2Baba  2004     1   39
1          2Baba  2006     1   34
2          2Baba  2010     1    0
3          2Baba  2011     1    0
4          2Baba  2014     2   39
..           ...   ...   ...  ...
327       Yung L  2013     1   14
328       Zlatan  2019     1   54
329        chike  2020     1   30
330  prettyboydo  2018     1   29
331  prettyboydo  2020     1   23

[332 rows x 4 columns]